overlord/snapstate/target.go

```diff
@@ -677,15 +683,53 @@ func (p *pathInstallGoal) toInstall(ctx context.Context, st *state.State, opts O
 		return nil, err
 	}
 
+	comps, err := componentSetupsFromPaths(info, p.components)
+	if err != nil {
+		return nil, err
+	}
+
 	inst := target{
 		setup: SnapSetup{
 			SnapPath:  p.path,
 			Channel:   channel,
 			CohortKey: p.revOpts.CohortKey,
 		},
-		info:   info,
-		snapst: snapst,
+		info:       info,
+		snapst:     snapst,
+		components: comps,
 	}
 
 	return []target{inst}, nil
 }
+
+func componentSetupsFromPaths(snapInfo *snap.Info, components map[*snap.ComponentSideInfo]string) ([]ComponentSetup, error) {
+	setups := make([]ComponentSetup, 0, len(components))
+	for csi, path := range components {
+		compInfo, err := validatedComponentInfo(path, csi, snapInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		setups = append(setups, ComponentSetup{
+			CompPath:     path,
+			CompSideInfo: csi,
+			CompType:     compInfo.Type,
+		})
+	}
+
+	return setups, nil
+}
+
+func validatedComponentInfo(path string, csi *snap.ComponentSideInfo, si *snap.Info) (*snap.ComponentInfo, error) {
```

Comment:
nitpick: maybe order of arguments should be the same as for `backend.OpenComponentFile`